### PR TITLE
Describe activation

### PR DESCRIPTION
**Pull request by VectorShadow**
Examining identified magic devices and other items that can be activated now presents an additional description of the odds of successful activation.

### CODE REVIEW

File: src/server/xtra1.c [hunk 1]
```diff
@@ -152,13 +168,7 @@ static void prt_level(int Ind)
...
 	player_type *p_ptr = Players[Ind];
 
-	int adv_exp;
-
-	if (p_ptr->lev >= PY_MAX_LEVEL)
-		adv_exp = 0;
-	else adv_exp = (s32b)(player_exp[p_ptr->lev - 1] * p_ptr->expfact / 100L);
-
-	Send_experience(Ind, p_ptr->lev, p_ptr->max_exp, p_ptr->exp, adv_exp);
+	send_indication(Ind, IN_LEVEL, MAX(p_ptr->max_plv, p_ptr->lev), p_ptr->lev);
```
Comment (evileye-uk) [inline src/server/xtra1.c]: This seems unnecessary, as with the similar cases below. Is there not a better way do do this? The giveaway is the fact that the parameters are repeated and the functions appear side-by-side.

File: src/server/xtra1.c [hunk 2]
```diff
@@ -402,8 +464,41 @@ static void prt_cut(int Ind)
 	player_type *p_ptr = Players[Ind];
 
 	int c = p_ptr->cut;
+	int s = 0;
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Look up table?

File: src/server/xtra1.c [hunk 3]
```diff
@@ -413,38 +508,206 @@ static void prt_stun(int Ind)
...
+	/* Blind */
+	if (p_ptr->blind)
+		cv_put_str(dest, TERM_ORANGE, "Blind", COL_BLIND, 5);
+
+	/* Confused */
+	if (p_ptr->confused)
+		cv_put_str(dest, TERM_ORANGE, "Confused", COL_CONFUSED, 8);
+
+	/* Afraid */
+	if (p_ptr->poisoned)
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Does poison make the player afraid, or just certain poison? Is this intentional?

File: src/server/xtra1.c [hunk 4]
```diff
@@ -531,14 +789,16 @@ static void prt_player_sust_info(int Ind)
...
 	player_type *p_ptr = Players[Ind];
 
-	int i, row, col, stat;
+	int i, row, col, stat, boost;
 	object_type *o_ptr;
-	u32b f1, f2, f3, f4;
-	u32b ignore_f2, ignore_f3, ignore_f4;
+	object_kind *k_ptr;
+	ego_item_type *e_ptr;
+	u32b f1, f2, f3, f1_hack = 0;
```
Comment (evileye-uk) [inline src/server/xtra1.c]: What is f1_hack?

File: src/server/xtra1.c [hunk 5]
```diff
@@ -520,8 +778,8 @@ static void prt_player_equippy(int Ind)
 		}
 
 		/* Dump proper character */
-		p_ptr->hist_flags[0][i-INVEN_WIELD].a = a;
```
Comment (evileye-uk) [inline src/server/xtra1.c]: What happened to the arrays?

File: src/server/xtra1.c [hunk 6]
```diff
@@ -716,8 +1001,8 @@ static void prt_player_flag_info(int Ind)
 					attr = TERM_WHITE;
 					c = '*';
 					//c_put_str(TERM_WHITE, "*", row, col+n);
-					//p_ptr->hist_flags[realY][realX].a = TERM_WHITE;
-					//p_ptr->hist_flags[realY][realX].c = '*';
+					//p_ptr->hist_flags[realX][realY].a = TERM_WHITE;
```
Comment (evileye-uk) [inline src/server/xtra1.c]: This now does a rotated nothing.

File: src/server/xtra1.c [hunk 7]
```diff
@@ -726,54 +1011,54 @@ static void prt_player_flag_info(int Ind)
 					//c_put_str(TERM_WHITE, "+", row, col+n);
 					attr = TERM_WHITE;
 					c = '+';
-					//p_ptr->hist_flags[realY][realX].a = TERM_WHITE;
-					//p_ptr->hist_flags[realY][realX].c = '+';
+					//p_ptr->hist_flags[realX][realY].a = TERM_WHITE;
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Why?

File: src/server/xtra1.c [hunk 8]
```diff
@@ -829,15 +1125,64 @@ static void prt_plusses(int Ind)
...
+	if (object_known_p(p_ptr, o_ptr)) show_todam += o_ptr->to_d;
 
-	Send_plusses(Ind, show_tohit, show_todam);
+	send_indication(Ind, IN_PLUSSES, show_tohit, show_todam);
 }
 
 static void prt_skills(int Ind)
 {
-	Send_skills(Ind);
+	player_type *p_ptr = Players[Ind];
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Why? This function is now long and complicated.

File: src/server/xtra1.c [hunk 9]
```diff
@@ -829,15 +1125,64 @@ static void prt_plusses(int Ind)
...
+		skills[0], factors[0],
+		skills[4], factors[4],
+		skills[1], factors[1],
+		skills[5], factors[5],
+		skills[2], factors[2],
+		skills[6], factors[6],
+		skills[3], factors[3],
+		skills[7], factors[7]);
+
+	send_indication(Ind, IN_SKILLS2,	skills[8], skills[9], skills[10]);
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Is this the reason it is long and complicated?

File: src/server/xtra1.c [hunk 10]
```diff
@@ -1235,11 +1522,13 @@ static void fix_spell(int Ind)
 	if (!p_ptr->cp_ptr->spell_book)
 		return;
 
+#if 0
```
Comment (evileye-uk) [inline src/server/xtra1.c]: aka delete

File: src/server/xtra1.c [hunk 11]
```diff
@@ -2128,25 +2402,25 @@ static void calc_bonuses(int Ind)
 	}
 
 
-	/* -APD- Hack -- rogues +1 speed at 5,20,35,50. this may be out of place, but.... */
-	
-	if (p_ptr->pclass == CLASS_ROGUE) 
+	/* Option : give (most!) rogues a bonus speed point on level 5,20,35,50 -APD- */
+	if ((p_ptr->cp_ptr->flags & CF_SPEED_BONUS) && !option_p(p_ptr,UNSETH_BONUS))
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Is this a repetition of line 706? Function?

File: src/server/xtra1.c [hunk 12]
```diff
@@ -2564,25 +2833,70 @@ static void calc_bonuses(int Ind)
...
+	p_ptr->skill_dev += adj_int_dev[p_ptr->stat_ind[A_INT]];
+
+	/* Affect Skill -- saving throw (WIS) */
+	p_ptr->skill_sav += adj_wis_sav[p_ptr->stat_ind[A_WIS]];
+
+	/* Affect Skill -- digging (STR) */
+	p_ptr->skill_dig += adj_str_dig[p_ptr->stat_ind[A_STR]];
+
+	/* Affect Skill -- disarming (Level, by Class) */
+	p_ptr->skill_dis += (p_ptr->cp_ptr->x_dis * p_ptr->lev / 10);
```
Comment (evileye-uk) [inline src/server/xtra1.c]: y += f(x)
repeat

File: src/server/xtra1.c [hunk 13]
```diff
@@ -2564,25 +2833,70 @@ static void calc_bonuses(int Ind)
...
+	p_ptr->skill_fos += (p_ptr->cp_ptr->x_fos * p_ptr->lev / 10);
+
+	/* Affect Skill -- combat (normal) (Level, by Class) */
+	p_ptr->skill_thn += (p_ptr->cp_ptr->x_thn * p_ptr->lev / 10);
+
+	/* Affect Skill -- combat (shooting) (Level, by Class) */
+	p_ptr->skill_thb += (p_ptr->cp_ptr->x_thb * p_ptr->lev / 10);
+
+	/* Affect Skill -- combat (throwing) (Level, by Class) */
+	p_ptr->skill_tht += (p_ptr->cp_ptr->x_thb * p_ptr->lev / 10);
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Wait. Copy - paste, or is this line correct?
tht += f(thb)?

File: src/server/xtra1.c [hunk 14]
```diff
@@ -2626,7 +2939,7 @@ static void calc_bonuses(int Ind)
 		}
 
 		/* Hack -- Reward High Level Rangers using Bows */
-        if ((p_ptr->pclass == CLASS_RANGER) && (p_ptr->tval_ammo == TV_ARROW))	
+       if ((p_ptr->cp_ptr->flags & CF_EXTRA_SHOT) && (p_ptr->tval_ammo == TV_ARROW))
```
Comment (evileye-uk) [inline src/server/xtra1.c]: The comment no longer makes sense. Needs better explanation.

File: src/server/xtra1.c [hunk 15]
```diff
@@ -2728,7 +3019,7 @@ static void calc_bonuses(int Ind)
 	p_ptr->icky_wield = FALSE;
 
 	/* Priest weapon penalty for non-blessed edged weapons */
-    if ((p_ptr->pclass == CLASS_PRIEST) && (!p_ptr->bless_blade) &&
+    if ((p_ptr->cp_ptr->flags & CF_BLESS_WEAPON) && (!p_ptr->bless_blade) &&
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Again, comment makes no sense anymore. Functions can be good for expressing intentions.

File: src/server/xtra1.c [hunk 16]
```diff
@@ -3129,16 +3386,19 @@ void redraw_stuff(int Ind)
 		cursor_redraw(Ind);
 	}
 
-	if (p_ptr->redraw & PR_HISTORY)
+	/*if (p_ptr->redraw & PR_HISTORY)
```
Comment (evileye-uk) [inline src/server/xtra1.c]: Delete

File: src/server/xtra2.c [hunk 17]
```diff
@@ -107,7 +116,7 @@ bool set_blind(int Ind, int v)
 	if (!notice) return (FALSE);
 
 	/* Disturb */
-	if (p_ptr->disturb_state) disturb(Ind, 0, 0);
+	if (option_p(p_ptr,DISTURB_STATE)) disturb(Ind, 0, 0);
```
Comment (evileye-uk) [inline src/server/xtra2.c]: This line is repeated a LOT, including the lack of space after the comma. Function? Look at the patterns the code makes. There are little blocks off stuff being done, separated by these disturb() calls. If only there were some way to group these logically.

File: src/server/xtra2.c [hunk 18]
```diff
@@ -1446,36 +1509,43 @@ bool set_cut(int Ind, int v)
 			/* Graze */
 			case 1:
 			msg_print(Ind, "You have been given a graze.");
+			sound(Ind, MSG_CUT);
```
Comment (evileye-uk) [inline src/server/xtra2.c]: I'm assuming the cut noise doesn't get louder based on how bad it is. How about we call it once?

File: src/server/xtra2.c [hunk 19]
```diff
@@ -1693,21 +1770,25 @@ bool set_food(int Ind, int v)
 			/* Weak */
 			case 1:
 			msg_print(Ind, "You are getting weak from hunger!");
+			sound(Ind, MSG_NOTICE);
```
Comment (evileye-uk) [inline src/server/xtra2.c]: Just checking this one.. NOTICE, right?

File: src/server/xtra2.c [hunk 20]
```diff
@@ -2270,140 +2367,224 @@ void monster_death(int Ind, int m_idx)
...
+			/* Found some one */
+	 		if (Ind != i)
+	 		{
+	 			count++;
+	 			found = i; 
+	 		}
+	 	}
+	}
+	if (count > 1) found = -2;
+#else
```
Comment (evileye-uk) [inline src/server/xtra2.c]: This code looks unused.

File: src/server/xtra2.c [hunk 21]
```diff
@@ -4084,506 +5112,2184 @@ int level_speed(int Ind)
...
+		p = "Magnificent";
+	}
+
+	/* Normal */
+	else
+	{
+		p = c_text + p_ptr->cp_ptr->title[(p_ptr->lev - 1) / 5];
+	}
+
+	center_string(buf, sizeof(buf), p_ptr->name);
```
Comment (evileye-uk) [inline src/server/xtra2.c]: You're about to see this a lot.. Function?

File: src/server/spells1.c [hunk 22]
```diff
@@ -1854,7 +1869,10 @@ static bool project_f(int Ind, int who, int r, int Depth, int y, int x, int dam,
 				}
 
 				/* Destroy the wall */
-				c_ptr->feat = FEAT_DIRT;
```
Comment (evileye-uk) [inline src/server/spells1.c]: Take a look at line 1889

File: src/server/cmd2.c [hunk 23]
```diff
@@ -979,514 +1154,1448 @@ static bool twall(int Ind, int y, int x)
...
+	cave_type		*c_ptr;
+	object_type		*o_ptr;
+
+	int y, x;
+
+	s16b o_idx;
+
+	bool more = FALSE;
+
+#if 0
```
Comment (evileye-uk) [inline src/server/cmd2.c]: delete

File: src/server/cmd2.c [hunk 24]
```diff
@@ -979,514 +1154,1448 @@ static bool twall(int Ind, int y, int x)
...
+	/* Door */
+	else
+	{
+		/* Open the door */
+		more = do_cmd_open_aux(Ind, y, x);
+	}
+
+	/* Cancel repeat unless we may continue */
+	if (!more) disturb(Ind, 0, 0);
+}
```
Comment (evileye-uk) [inline src/server/cmd2.c]: Out of interest, how many lines long is this function?

File: src/server/cmd2.c [hunk 25]
```diff
@@ -979,514 +1154,1448 @@ static bool twall(int Ind, int y, int x)
...
+	/* Door */
+	else
+	{
+		/* Close door */
+		more = do_cmd_close_aux(Ind, y, x);
+	}
+
+	/* Cancel repeat unless told not to */
+	if (!more) disturb(Ind, 0, 0);
+}
```
Comment (evileye-uk) [inline src/server/cmd2.c]: Nearly 100 lines to close a door? There has to be a better way.

File: src/server/cmd2.c [hunk 26]
```diff
@@ -979,514 +1154,1448 @@ static bool twall(int Ind, int y, int x)
...
+		{
+			/* We may continue tunelling */
+			msg_print(Ind, "You tunnel into the door.");
+			more = TRUE;
+		}
+	}
+
+	/* Result */
+	return (more);
+}
```
Comment (evileye-uk) [inline src/server/cmd2.c]: I was lost at line 1900

File: src/server/cmd2.c [hunk 27]
```diff
@@ -1502,156 +2611,234 @@ void do_cmd_disarm(int Ind, int dir)
...
+	/* Get grid */
+	c_ptr = &cave[Depth][y][x];
+
 	/* Allow repeated command */
-	if (command_arg)
+	if (p_ptr->command_arg)
 	{
 		/* Set repeat count */
-		/*command_rep = command_arg - 1;*/
+		/*p_ptr->command_rep = p_ptr->command_arg - 1;*/
```
Comment (evileye-uk) [inline src/server/cmd2.c]: What does this line do?

File: src/server/cmd2.c [hunk 28]
```diff
@@ -1899,29 +3107,15 @@ int do_cmd_run(int Ind, int dir)
...
-		 * try to queue the run command.
-		 */
-		else return 0;
+		/* Initialise running */
+		p_ptr->run_request = dir;
+		p_ptr->running = TRUE;
+		p_ptr->ran_tiles = 0;
 	}
-	return 2;
+	return 1;
```
Comment (evileye-uk) [inline src/server/cmd2.c]: Is 1 better or worse than 2?

File: src/server/cmd2.c [hunk 29]
```diff
@@ -2393,17 +3626,20 @@ void do_cmd_fire(int Ind, int dir, int item)
 				dispx = x - p_ptr->panel_col_prt;
 
 				/* Remember the projectile */
-				p_ptr->scr_info[dispy][dispx].c = missile_char;
```
Comment (evileye-uk) [inline src/server/cmd2.c]: All this looks unfinished.

File: src/server/cmd2.c [hunk 30]
```diff
@@ -2793,17 +4060,20 @@ void do_cmd_throw(int Ind, int dir, int item)
...
-				p_ptr->scr_info[dispy][dispx].a = missile_attr;
+				//p_ptr->scr_info[dispy][dispx].c = missile_char;
```
Comment (evileye-uk) [inline src/server/cmd2.c]: This is/appears to be a duplicate of line 3629.

File: src/server/cmd4.c [hunk 31]
```diff
@@ -579,47 +727,522 @@ void do_cmd_check_players(int Ind, int line)
...
+		
+		/* Ask for a choice */
+		text_out(" \n");
+		text_out(" \n");
+		text_out(" \n");
+		text_out("Display current knowledge\n");
+		text_out(" \n");
+
+		/* Give some choices */
+		text_out("    (1) Display known artifacts\n");
```
Comment (evileye-uk) [inline src/server/cmd4.c]: This is a server file, right?

File: src/server/generate.c [hunk 32]
```diff
@@ -477,7 +477,7 @@ static void place_locked_door(int Depth, int y, int x)
 	cave_type *c_ptr = &cave[Depth][y][x];
 
 	/* Create locked door */
-	c_ptr->feat = FEAT_DOOR_HEAD + randint(7);
```
Comment (evileye-uk) [inline src/server/generate.c]: Starting to see repeating patterns again...

**Comment by VectorShadow** [comment]
Hey, sorry, I’ve been working with Flambard on this – unfortunately, I’m not extremely experienced with git, so I believe I may have been editing the wrong branch entirely, somehow.

The only change I made was adding a new method in obj-info.c to describe the odds of successfully activating a magic device or other item.

I’ll retrace my steps and see if I can find out where I went wrong - I'm not attempting to edit the development branch at all.

Apologies,
-Avenger

**Comment by evileye-uk** [comment]
Put in a new PR with the right branch, then. That does look suspiciously like a merge from master. Switch branches over to develop and cherry-pick the bits you want. Some of that stuff on master really shouldn't see the light of day.

**Comment by evileye-uk** [comment]
You can back track just by switching branches.. Do a checkout of develop, then fetch & pull/merge it. Create new branch, make the changes you want; add them, commit & push. When you view the new PR, it should have one file changed, rather than many.

**Comment by evileye-uk** [comment]
Closing this.

**Comment by VectorShadow** [comment]
Yes, checking it on my end, it appears to be exactly master, which explains why I wasn't able to submit a pull request to master, because there are no changes. I had intended to push my local changes to my GitHub repository, but apparently that was not successful. I'll find out what I'm doing wrong, and I should be back with a very small commit to master, not develop. I appreciate your feedback.

**Pull request closed without merging**